std/socket.zig

```diff
@@ -0,0 +1,678 @@
+                    0 => return initSocket(@intCast(SocketFd, rc), self.domain, self.socket_type, self.protocol),
+                    system.EINTR => continue,
+                    else => return unexpectedError(err),
+                    system.EAGAIN => unreachable, // use asyncAccept for non-blocking
+                    system.EBADF => unreachable, // always a race condition
+                    system.ECONNABORTED => return AcceptError.ConnectionAborted,
+                    system.EFAULT => unreachable,
+                    system.EINVAL => unreachable,
+                    system.EMFILE => return AcceptError.ProcessFdQuotaExceeded,
+                    system.ENFILE => return AcceptError.system.emFdQuotaExceeded,
+                    system.ENOBUFS => return AcceptError.system.emResources,
+                    system.ENOMEM => return AcceptError.system.emResources,
+                    system.ENOTSOCK => return AcceptError.FileDescriptorNotASocket,
+                    system.EOPNOTSUPP => return AcceptError.OperationNotSupported,
+                    system.EPROTO => return AcceptError.ProtocolFailure,
+                    system.EPERM => return AcceptError.BlockedByFirewall,
+                }
+            }
+        } else {
+            while (true) {
+                var sockaddr_size = system.socklen_t(@sizeOf(net.OsAddress));
+                // port accept4
+                const rc = accept4Windows(self.fd, &addr.os_addr, &sockaddr_size, flags);
+                const err = system.WSAGetLastError();
+                switch (err) {
+                    0 => {
+                        if (rc == InvalidSocketFd) {
+                            return unexpectedError(@intCast(u32, system.GetLastError()));
+                        } else {
+                            return initSocket(rc, self.domain, self.socket_type, self.protocol);
+                        }
+                    },
+                    system.ERROR.WSAEINTR => continue,
+                    else => return unexpectedError(@intCast(u32, err)),
+                    // TODO check for TOO_MANY_OPEN_FILES?
+                    system.ERROR.WSAEWOULDBLOCK => unreachable, // use asyncAccept for non-blocking
+                    system.ERROR.WSAECONNRESET => return AcceptError.ConnectionAborted,
+                    system.ERROR.WSAEFAULT => unreachable,
+                    system.ERROR.WSAEINVAL => unreachable,
+                    system.ERROR.WSAEMFILE => return AcceptError.ProcessFdQuotaExceeded,
+                    system.ERROR.WSAENOBUFS => return AcceptError.system.emResources,
+                    system.ERROR.WSAENOTSOCK => return AcceptError.FileDescriptorNotASocket,
+                    system.ERROR.WSAEOPNOTSUPP => return AcceptError.OperationNotSupported,
+                }
+            }
+        }
+    }
+
+    /// Returns InvalidSocketFd if would block.
+    pub fn asyncAccept(self: Socket, addr: *net.Address, flags: u32) AcceptError!Socket {
```

Comment:
I don't like having these async variants when it relies on the kernel-side socket state.

Reply:
these are taken from `std.os`.